### ISSUE TITLE
mosh with background processes doesn't disconnect

### ISSUE THREAD

**Issue by unode** [mobile-shell/mosh#1035]
Steps to reproduce:

1. `mosh server bash`
2. `sleep 60 &`
3. `disown`
4. `Ctrl+D`

At step 4 I would expect mosh to disconnect, however it only disconnects when `sleep 60` finishes.

Versions: `mosh 1.3.2 (client)` - `mosh 1.3.0 (server)`

**Comment by cgull** [comment]
Can you better explain the larger problem you're having here?  Given your repro, you appear to be trying to run `mosh-server` without using the `mosh` script.  We do try to support that, but everyone's situation is a little different for this.

Also, `mosh-server` detaches itself from the terminal.  It must, because the initiating session may terminate before it establishes connection with the client.  And coincidentally, the timeout after not hearing anything from the client is 60 seconds.  So I think `mosh-server` is exiting at 60 seconds, but not for the reason you think it's exiting.

**Comment by unode** [comment]
Sorry for being too brief.

The test-case above was meant for reproducing the issue and doesn't detail the original problem.

My typical use and the reason that motivated this issue is slightly more elaborate.
I often login to a remote server, run a command with (or without) `nohup` and then disown to ensure it doesn't get killed when I disconnect.

If I do this using `ssh` everything works as expected (job remains and connection closes as soon as I exit). However with  `mosh server.name`, I `exit` the remote shell, the terminal clears the output, but doesn't exit `mosh` (terminal remains busy). This doesn't happen if there is no background or disowned job on the remote server. 

I haven't tested the 60 seconds timeout (I don't usually wait that long) but so far it seems that the connection is only closed when the `disown`'ed command on the remote server ends. At least, on tests with increasing sleep times the server disconnects when the `sleep` command finishes.

**Comment by cgull** [comment]
Ah, I think I misunderstood your original report.  You're running steps 2-4 in the newly created Mosh terminal session, not in the initial SSH session, correct?  If so, ignore my initial guesses, they're wrong.

Assuming that's true, I can't reproduce the problem on macOS or FreeBSD 12.0, but I can on openSUSE.  That's a slightly odd problem, we'll take a look.

**Comment by unode** [comment]
>  You're running steps 2-4 in the newly created Mosh terminal session,

Correct.

I can reproduce the issue when the server is CentOS 6 and 7 (mosh 1.3.0) and Ubuntu 16.04 (mosh 1.2.5) don't have access to FreeBSD to test. Client OS doesn't seem to make a difference, tried NixOS, Gentoo and Ubuntu (mosh 1.3.2) and all showed the same behavior.

**Comment by petesh** [comment]
> This also happens on a stock standard ssh session as well - I've always assumed that even after disowning jobs I'd end up having to force terminate the shell parent because of the pty being held open

This is relatively easy to reproduce at least on ubuntu 18.10

the problem is that the subprocess that's been disowned has a handle open to the pty.

So you have:
mosh -> (shell) -> subprocess

the (shell) has terminated, and is exited, but the subprocess is keeping open the ptys for stdin, stdout and stderr.

**Comment by plup** [comment]
I'm having the issue as well on Debian buster (client & server).

**Comment by unode** [comment]
Indeed, I can confirm that the original report works if using `nohup sleep 60 &`.
However the issue remains if any background processes have open ptys as mentioned by @petesh .
